### ISSUE TITLE
Stop suggesting using unwrap to have inline result

### ISSUE THREAD

**Issue by Luk-z** [reduxjs/redux-toolkit#3515]
Imo these 2 statements are wrong and encourage developers using old-style Promise syntax instead `async/await`.

`If you need to access the error or success payload immediately after a mutation, you can chain .unwrap().`
https://redux-toolkit.js.org/rtk-query/usage/error-handling#error-display-examples

`This can be useful if you wish to determine whether the mutation succeeds/fails inline at the call-site.`
https://redux-toolkit.js.org/rtk-query/usage/mutations#frequently-used-mutation-hook-return-values

If i need to access payload immediately after a mutation I will use

```
const [trigger] = useMutation();
...
...
const result = await trigger(...);
// result is "the error or success payload retrieved immediately after a mutation" and can be used "to determine whether the mutation succeeds/fails inline at the call-site"
```

**Comment by EskiMojo14** [comment]
i think you're missing what unwrap does.
By default, the promise returned by the trigger will *always* resolve, with either `{ data }` or `{ error }`.
Calling .unwrap on that promise returns a *new* promise, that will either resolve with data, or reject with error.

None of this has anything to do with "old style Promise syntax" vs async/await, you can use either.
```ts
const resultPromise = trigger();
const result = await resultPromise; // { data } or { error }
try {
  const data = await resultPromise.unwrap();
} catch (error) {
  console.error(error);
}
```

**Comment by Luk-z** [comment]
ok, let's forget about `and encourage developers using old-style Promise syntax instead async/await.`, effectively it's wrong, my bad.

The point is:
`If you need to access the error or success payload immediately after a mutation, you can chain .unwrap().` is confusing because encourage developers to use `unwrap()` instead the wrapped result (using a common interface). 
I think a better sentence will be:
`If you need to access the error or success payload immediately after a mutation, you can await the trigger result or chain .unwrap().` or something similar.

Also `This can be useful if you wish to determine whether the mutation succeeds/fails inline at the call-site.` is wrong because `to determine whether the mutation succeeds/fails inline at the call-site` it's better using the wrapped result.

The reason because it's better using the wrapped result is to have a common interface that defines the result.

**Comment by EskiMojo14** [comment]
that seems like a matter of opinion - one we may not actually agree with.

for example if you're using typescript, in order to tell whether the "wrapped" result is an error or not, you would have to do the below:
```ts
const result = await trigger();
if ('error' in result) {
  console.error(result.error);
} else {
  const { data } = result;
}
``` 
Compare this to .unwrap, which allows you to handle it in the same way you'd normally handle a rejected request: 
```ts
try {
  const data = await trigger().unwrap();
} catch (error) {
  console.error(error)
}
```

**Comment by phryneas** [comment]
I'm not gonna go into the "FP-style vs Imperative programming" discussion here. Generally, Promises are a thing, and they have a mechanism for errors: rejecting the promise.

So the question is why we need `unwrap` in the first place, and deviate from that "official" mechanism. In our case, we wanted to make sure that at no point in time, an "uncaught promise rejection" bubbles up because someone did not `.catch` - so we introduced a "never-throwing" promise for that purpose only.  
Later we added `.unwrap` for it, essentially for everyone who actually wants to await (and catch) that data.

If I were redesigning the api nowadays, I would go with a rejecting promise from the start, but one that would be caught in parallel, and not cause the uncaught promise rejection. It's too late for that though.

As for the typesafety of the error: That error will be a union type anyways, so no matter if you catch it, or if you have a shared interface, you will need some type of guard function. No way of writing this is more useful than the other here.

**Comment by sebastien-f** [comment]
While I do understand the motivation behind the "never-throwing promise", I think that's an enormous design error that encourages bad practices and void the whole purpose of it. It also means that redux-toolkit behaves in an unexpected way. 

Beyond that, while I can imagine why you'd want to have an union-ed error type for practical reasons, it prevents any use of `error instanceof X` to handle it properly and force to rely on other type detection.

Not to mention that if using Sentry (which made me look into this behavior), the error is regarded as non Error and thus treated differently. That's one product, but there might be others that suffers from the same issue.

**Comment by zernie** [comment]
It'd be nice to have a config that disables this behavior. it's pretty counter-intuitive.

**Comment by EskiMojo14** [comment]
@zernie I don't think that would be possible to type correctly.

**Comment by zernie** [comment]
Sorry, why?

**Comment by EskiMojo14** [comment]
could you give an example of a potential API for the config?

**Comment by zernie** [comment]
Tbh I thought mutate() and mutate().unwrap both return a promise so it shouldn't affect the types. Am I wrong?

**Comment by EskiMojo14** [comment]
they're both promises, but promises that resolve to different values. mutate() resolves to `{ data } | { error }`, and mutate().unwrap() resolves to `data` *or* rejects with `error`.
